### PR TITLE
Add a `builtin_type_shadow` lint

### LINKED ISSUE #949: shadowing primitive types
It's legal in beta/nightly to shadow primitive types. I don't think this is useful to have, but it was decided to do this for forward compatibility with `u128` and potentially other new primitive types. Clippy could lint against any type declarations or generic arguments that shadow a primitive

### PR DESCRIPTION
**Pull request by mcarton**
See https://github.com/rust-lang/rust/issues/36046.
Close #949.

### CODE REVIEW

File: clippy_lints/src/misc_early.rs [hunk 1]
```diff
@@ -141,6 +141,27 @@ declare_lint! {
     "integer literals starting with `0`"
 }
 
+/// **What it does:** Warns if a generic shadows a built-in type.
+///
+/// **Why is this bad?** This gives surprising type error.
```
Comment (killercup) [inline clippy_lints/src/misc_early.rs]: Either 'errors' or 'a surprising type error'

File: clippy_lints/src/utils/constants.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+//! This module contains some useful constants.
+
+#![deny(missing_docs_in_private_items)]
+
+/// List of the built-in types names.
```
Comment (killercup) [inline clippy_lints/src/utils/constants.rs]: Maybe link to https://doc.rust-lang.org/reference.html#primitive-types as a source?

**Comment by mcarton** [comment]
Rebased and addressed comments.

**Comment by killercup** [comment]
LGTM

I was confused by the wording ("shadowing") a bit, bit it makes sense. (Maybe we can add a span label for the generic name later on, saying something like "The generic `{}` has the same name as a built-in type", just to also have alternative wording.)

**Comment by llogiq** [comment]
Good lint. @Manishearth we should propose adding this to rustc. It's really a footgun.

**Comment by Manishearth** [comment]
Not as a lint, but that error message could get a help message

**Pull request merged**